aten/src/ATen/native/TensorDimApply.h

```diff
@@ -0,0 +1,52 @@
+#include <ATen/ATen.h>
+
+namespace at {
+  namespace native {
+    //input tensors are non-zero dim and non-empty
+    template<typename T1, typename T2, typename Operation, typename Function>
+    void tensor_dim_apply3(const Tensor& self, Tensor& values, Tensor& indices, int64_t dim, Operation op, Function func) {
```

Comment:
This is very awkward when you have both op and func as your arguments, please restructure your code so that you need only one of them.